### PR TITLE
[torch.compile]: Optimize collectives in TP MoE case using torch.compile pass

### LINKED ISSUE #29139: [Feature]: Optimize collectives in TP MoE case using torch.compile pass
### 🚀 The feature, motivation and pitch

To avoid redundant work in MoE models in the TP case, sequence parallelism was added to the Deepseek model definition in #24134 and expanded to other models in #24982. However, to avoid performing surgery on the linear layer, the current approach performs more communication than necessary. With a torch.compile custom pass, we can rewrite the graph to remove the redundant computation.

### More details

Before the SP optimization, the ops in the model were:
```
- o_proj:[num_tokens, ...] -> [num_tokens, ...] (incomplete results)
- all_reduce:[num_tokens, ...] -> [num_tokens, ...]
- rms_norm:[num_tokens, ...] -> [num_tokens, ...]
- router:[num_tokens, ...] -> [num_tokens, ...]
- experts:[num_tokens, ...] -> [num_tokens, ...]
- ...
```

With sequence parallel enabled, this becomes:
```
- o_proj: [num_tokens, ...] -> [num_tokens, ...] (incomplete results)
- all_reduce: [num_tokens, ...] -> [num_tokens, ...]
- rms_norm:[num_tokens, ...] -> [num_tokens, ...]
- chunk: [num_tokens, ...] -> [num_tokens/tp, ...]
- router: [num_tokens/tp, ...] -> [num_tokens/tp, ...]
- experts: [num_tokens/tp, ...] -> [num_tokens/tp, ...]
- all_gather: [num_tokens/tp, ...] -> [num_tokens, ...]
```

Additionally, experts now properly do the dp+tp<->ep dispatch instead of just the original replicated dp<->ep dispatch.

Notice that the `all_reduce` (and `rms_norm`) do redundant communication as each TP rank only requires partial results. With a compile pass, we can convert the `all_reduce` -> `rms_norm` -> `chunk` sequence into a `reduce_scatter` -> `rms_norm`:

```
- o_proj: [num_tokens, ...] -> [num_tokens, ...] (incomplete results)
- reduce_scatter: [num_tokens, ...] -> [num_tokens/tp, ...]
- router: [num_tokens/tp, ...] -> [num_tokens/tp, ...]
- experts: [num_tokens/tp, ...] -> [num_tokens/tp, ...]
- all_gather: [num_tokens/tp, ...] -> [num_tokens, ...]
```

We should create a new `SequenceParallelismMoEPass`, controlled by a new `PassConfig.enable_sp_moe` flag (following the new naming convention in #27995) so that it can be turned on independently of regular SP. We will likely need to pad the number of tokens to a multiple of TP size, although like described in #29136, there are alternatives.

#### SP
Additionally, to make SP-MoE compatible with AsyncTP, the SP pass should also replace the `all_gather` -> `rms_norm` sequence with `rms_norm` -> `all_gather`, such that the `all_gather` can be fused with the following QKV proj. We can see that in the SP case we perform an `all_gather` instead of `all_reduce` after the MLP layer.

### Alternatives

Alternatively, the original optimization could be done as a compile pass as well, which would significantly clean up the MoE model definitions. However, that would mean that `VLLM_COMPILE` compilation mode would be required for this optimization and if compilation is disabled, the optimization would be disabled as well. Generally we accept lower performance in eager mode as compilation is on by default, but I know there was a reason this was done this way (don't remember why).

We could also just let the SP pass first transform `all_reduce` -> `rms_norm` into `reduce_scatter` -> `rms_norm` -> `all_gather` and then eliminate the `all_gather` -> `chunk` combo. But then we have less control over SP after MoE (although it seems like all_reduce is not even used in that case).

### Additional context

Original proposal comment: https://github.com/vllm-project/vllm/pull/24982#pullrequestreview-3259494618

cc @tlrmchlsmth @bnellnm @robertgshaw2-redhat @alexm-redhat @zou3519 @nvpohanh @youkaichao 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by SouthWest7**
## Purpose
This PR Resolves #29139 by optimizing communication patterns in TP MoE sequence-parallel
Replace the `all_reduce` at the end of `attn` with a `reduce_scatter`.
## Test Plan
- ut
```shell
pytest tests/compile/passes/distributed/test_sequence_parallelism_moe.py
```

`vllm serve`
```shell
CUDA_VISIBLE_DEVICES=0,1,2,3 vllm serve /data1/models/Qwen3-30B-A3B \
  --host 0.0.0.0 \
  --port 8000 \
  --tensor-parallel-size 2 \
  --data-parallel-size 2 \
  --enable-expert-parallel \
  --all2all-backend allgather_reducescatter \
  --trust-remote-code \
  --compilation-config '{"mode":3,"use_inductor_graph_partition":true,"splitting_ops":[],"pass_config":{"enable_sp":false,"enable_sp_moe":true}}'
```
- lm_eval
```shell
lm_eval \
  --model local-completions \
  --model_args model=/data1/models/Qwen3-30B-A3B,base_url=http://127.0.0.1:8000/v1/completions,num_concurrent=32,tokenized_requests=False \
  --tasks gsm8k \
  --batch_size auto
```

- e2e
```shell
vllm bench serve \
   --backend vllm \
   --model /data1/models/Qwen3-30B-A3B \
   --endpoint /v1/completions \
   --dataset-name sharegpt \
   --dataset-path /data1/datasets/ShareGPT_V3_unfiltered_cleaned_split.json \
   --num-prompts 100
```
## Test Result
- ut
```
====================================================== test session starts =======================================================
platform linux -- Python 3.12.12, pytest-9.0.2, pluggy-1.6.0
rootdir: /root/autodl-tmp/vllm
configfile: pyproject.toml
plugins: anyio-4.12.1, asyncio-1.3.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 20 items                                                                                                               

tests/compile/passes/distributed/test_sequence_parallelism_moe.py ....................                                     [100%]
```
- lm_eval
```shell
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8575|±  |0.0096|
|     |       |strict-match    |     5|exact_match|↑  |0.8992|±  |0.0083|
```

- e2e
`enable`
```shell
Starting initial single prompt test run...
Skipping endpoint ready check.
Starting main benchmark run...
Traffic request rate: inf
Burstiness factor: 1.0 (Poisson process)
Maximum request concurrency: None
100%|███████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:32<00:00,  3.07it/s]
tip: install termplotlib and gnuplot to plot the metrics
============ Serving Benchmark Result ============
Successful requests:                     100       
Failed requests:                         0         
Benchmark duration (s):                  32.53     
Total input tokens:                      23260     
Total generated tokens:                  22061     
Request throughput (req/s):              3.07      
Output token throughput (tok/s):         678.14    
Peak output token throughput (tok/s):    1934.00   
Peak concurrent requests:                100.00    
Total token throughput (tok/s):          1393.13   
---------------Time to First Token----------------
Mean TTFT (ms):                          8249.77   
Median TTFT (ms):                        8525.01   
P99 TTFT (ms):                           9468.02   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          68.07     
Median TPOT (ms):                        44.44     
P99 TPOT (ms):                           294.84    
---------------Inter-token Latency----------------
Mean ITL (ms):                           40.71     
Median ITL (ms):                         37.13     
P99 ITL (ms):                            342.32    
==================================================
```
```disable
Starting initial single prompt test run...
Skipping endpoint ready check.
Starting main benchmark run...
Traffic request rate: inf
Burstiness factor: 1.0 (Poisson process)
Maximum request concurrency: None
100%|███████████████████████████████████████████████████████████████████████████████████████████| 100/100 [00:31<00:00,  3.21it/s]
tip: install termplotlib and gnuplot to plot the metrics
============ Serving Benchmark Result ============
Successful requests:                     100       
Failed requests:                         0         
Benchmark duration (s):                  31.13     
Total input tokens:                      23260     
Total generated tokens:                  22061     
Request throughput (req/s):              3.21      
Output token throughput (tok/s):         708.72    
Peak output token throughput (tok/s):    2103.00   
Peak concurrent requests:                100.00    
Total token throughput (tok/s):          1455.95   
---------------Time to First Token----------------
Mean TTFT (ms):                          7364.03   
Median TTFT (ms):                        8927.47   
P99 TTFT (ms):                           9750.00   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          108.67    
Median TPOT (ms):                        41.79     
P99 TPOT (ms):                           1303.13   
---------------Inter-token Latency----------------
Mean ITL (ms):                           43.80     
Median ITL (ms):                         34.38     
P99 ITL (ms):                            135.70    
==================================================
```
---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/compilation/passes/fusion/sequence_parallelism_moe.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        )
+
+        pattern_dtype = self.model_dtype or torch.get_default_dtype()
+        # Register model eps first to reduce pattern-matching search space.
+        eps_candidates = [1e-5, 1e-6]
+        model_config = getattr(config, "model_config", None)
+        hf_text_config = getattr(model_config, "hf_text_config", None)
+        model_eps = getattr(hf_text_config, "rms_norm_eps", None)
+        if isinstance(model_eps, float):
+            eps_candidates = [model_eps, *[e for e in eps_candidates if e != model_eps]]
```
Comment (ProExpertProg) [inline vllm/compilation/passes/fusion/sequence_parallelism_moe.py]:
```suggestion
            eps_candidates = {model_eps, *eps_candidates}
```

File: vllm/distributed/parallel_state.py [hunk 2]
```diff
@@ -143,6 +144,39 @@ def reduce_scatter(
...
+    if dim < 0:
+        dim += tensor.dim()
+    assert 0 <= dim < tensor.dim(), (
+        f"Invalid dim ({dim}) for input tensor with shape {tensor.size()}"
+    )
+
+    remainder = tensor.shape[dim] % world_size
+    if remainder != 0:
+        pad_len = world_size - remainder
+        x = tensor.movedim(dim, 0)
```
Comment (ProExpertProg) [inline vllm/distributed/parallel_state.py]: Use `torch.nn.functional.pad()`?

File: vllm/compilation/passes/pass_manager.py [hunk 3]
```diff
@@ -116,6 +117,12 @@ def configure(self, config: VllmConfig) -> None:
             if self.pass_config.eliminate_noops:
                 self.passes += [NoOpEliminationPass(config)]
 
+            if (
+                self.pass_config.enable_sp_moe
+                and config.parallel_config.use_sequence_parallel_moe
+            ):
```
Comment (ProExpertProg) [inline vllm/compilation/passes/pass_manager.py]: If `config.parallel_config.use_sequence_parallel_moe == False`, this pass was already disabled in config init, no?
```suggestion
            if self.pass_config.enable_sp_moe:
```

File: vllm/compilation/passes/fusion/sequence_parallelism_moe.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    def is_applicable_for_range(self, compile_range: Range) -> bool:
+        # Align with SequenceParallelismPass behavior.
+        if (
+            not self.compilation_config.splitting_ops
+            or self.compilation_config.use_inductor_graph_partition
+        ):
+            return True
+        tp_size = get_tensor_model_parallel_world_size()
+        result: bool = (compile_range.is_single_size()) and (
+            compile_range.end % tp_size == 0
```
[the remark this answers is not in the document]
Comment (ProExpertProg) [reply vllm/compilation/passes/fusion/sequence_parallelism_moe.py]: I think this is true, no?

File: vllm/compilation/passes/fusion/sequence_parallelism_moe.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        # Fast-exit to avoid pattern matcher overhead for graphs without
+        # sequence_parallel_chunk.
+        chunk_op = torch.ops.vllm.sequence_parallel_chunk_impl.default
+        has_chunk = any(
+            node.op == "call_function" and node.target == chunk_op
+            for node in graph.nodes
+        )
+        if not has_chunk:
+            self.matched_count = 0
+            return
```
Comment (ProExpertProg) [inline vllm/compilation/passes/fusion/sequence_parallelism_moe.py]: I think let's just remove this, the pattern matcher should be just as fast (it just walks through the graph looking for this function just like your early exit code)

File: vllm/compilation/passes/fusion/sequence_parallelism_moe.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            residual_chunk = self._sequence_parallel_chunk(residual)
+            rmsnorm, residual_out = self.rmsnorm_matcher(
+                reduce_scatter,
+                weight,
+                residual_chunk,
+            )
+            # Keep residual interface full-sized for decoder-layer boundaries.
+            residual_full = self._all_gather(residual_out)
+            residual_full = residual_full[: residual.size(0), ...]
+            return rmsnorm, residual_full
```
Comment (ProExpertProg) [inline vllm/compilation/passes/fusion/sequence_parallelism_moe.py]: It's not clear to me that this is correct. How exactly does this work? You reconstruct the residual and then just take the first slice?

File: tests/compile/passes/distributed/test_sequence_parallelism_moe.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+
+@multi_gpu_test(num_gpus=2)
+@pytest.mark.parametrize("with_residual", [False, True])
+# Cover both divisible and non-divisible sequence lengths. The pass handles
+# odd lengths via reduce_scatter_with_padding.
+@pytest.mark.parametrize("seq_len", [15, 16])
+@pytest.mark.parametrize("hidden_size", [32])
+@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
+@pytest.mark.skipif(envs.VLLM_TARGET_DEVICE not in ["cuda"], reason="Only on CUDA")
```
Comment (ProExpertProg) [inline tests/compile/passes/distributed/test_sequence_parallelism_moe.py]: Can you also parametrize on rms_norm enabled/disabled?

**Review by ProExpertProg — COMMENTED** [review]
Overall structure looks great! Just implementation detail questions. Also please push a commit with @anant37289 to credit them for their initial work. And add this pass to the fusion E2E tests (tp2 Ar-rms and tp2 asynctp)

**Comment by ProExpertProg** [comment]
And please report some E2E perf numbers, perhaps for Qwen3 TP=4?

File: vllm/compilation/passes/fusion/sequence_parallelism_moe.py (hunk 6 above)
Comment (SouthWest7) [reply vllm/compilation/passes/fusion/sequence_parallelism_moe.py]: My understanding is that `SP` here is only applied to the MoE `hidden_states` path.
The decoder residual sits at the layer boundary, and it’s expected to stay **full-sized** across layers, since it’s passed into the next layer’s `fused add + RMSNorm`. 
In the replacement, I chunk the residual to match the local `reduce_scatter` computation. 
After that, I all-gather it back so the layer interface remains unchanged.
The final trim just removes the tail padding rows that were introduced during chunking when the sequence length isn’t divisible by the parallel size. It doesn’t drop any real tokens.
But this does introduce extra communication, which is not aligned with our original goal of minimizing communication overhead. So I’m quite confused about this.

Comment (ProExpertProg) [reply vllm/compilation/passes/fusion/sequence_parallelism_moe.py]: I see, that makes sense. In the current SP pass we just assume that this is applied to all rms norm calls so they all become sequence parallel, but I suppose you can't make that assumption because we're not converting all rmsnorm calls to SP.

Keep the current all gather + slice approach! And then I'm a follow-up let's add a pass that runs after SP and SP MoE and removes the all gathers if possible!

cc @jasonlizhengjian I think this might actually be a real issue with running SP on MoE models because not all rmsnorms are getting replaced
